user.go

```diff
@@ -348,124 +348,144 @@ func (u *user) run() {
-				uc.updateAway()
-			})
+				u.downstreamConns = append(u.downstreamConns, dc)
 
-			dc.updateSupportedCaps()
-		case eventDownstreamDisconnected:
-			dc := e.dc
+				u.forEachUpstream(func(uc *upstreamConn) {
+					uc.updateAway()
+				})
 
-			for i := range u.downstreamConns {
-				if u.downstreamConns[i] == dc {
-					u.downstreamConns = append(u.downstreamConns[:i], u.downstreamConns[i+1:]...)
-					break
+				dc.updateSupportedCaps()
+			case eventDownstreamDisconnected:
+				dc := e.dc
+
+				for i := range u.downstreamConns {
+					if u.downstreamConns[i] == dc {
+						u.downstreamConns = append(u.downstreamConns[:i], u.downstreamConns[i+1:]...)
+						break
+					}
 				}
-			}
 
-			// Save history if we're the last client with this name
-			skipHistory := make(map[*network]bool)
-			u.forEachDownstream(func(conn *downstreamConn) {
-				if dc.clientName == conn.clientName {
-					skipHistory[conn.network] = true
-				}
-			})
+				// Save history if we're the last client with this name
+				skipHistory := make(map[*network]bool)
+				u.forEachDownstream(func(conn *downstreamConn) {
+					if dc.clientName == conn.clientName {
+						skipHistory[conn.network] = true
+					}
+				})
 
-			dc.forEachNetwork(func(net *network) {
-				if skipHistory[net] || skipHistory[nil] {
-					return
-				}
+				dc.forEachNetwork(func(net *network) {
+					if skipHistory[net] || skipHistory[nil] {
+						return
+					}
 
-				net.offlineClients[dc.clientName] = struct{}{}
-			})
+					net.offlineClients[dc.clientName] = struct{}{}
+				})
 
-			u.forEachUpstream(func(uc *upstreamConn) {
-				uc.updateAway()
-			})
-		case eventDownstreamMessage:
-			msg, dc := e.msg, e.dc
-			if dc.isClosed() {
-				dc.logger.Printf("ignoring message on closed connection: %v", msg)
-				break
-			}
-			err := dc.handleMessage(msg)
-			if ircErr, ok := err.(ircError); ok {
-				ircErr.Message.Prefix = dc.srv.prefix()
-				dc.SendMessage(ircErr.Message)
-			} else if err != nil {
-				dc.logger.Printf("failed to handle message %q: %v", msg, err)
-				dc.Close()
+				u.forEachUpstream(func(uc *upstreamConn) {
+					uc.updateAway()
+				})
+			case eventDownstreamMessage:
+				msg, dc := e.msg, e.dc
+				if dc.isClosed() {
+					dc.logger.Printf("ignoring message on closed connection: %v", msg)
+					break
+				}
+				err := dc.handleMessage(msg)
+				if ircErr, ok := err.(ircError); ok {
+					ircErr.Message.Prefix = dc.srv.prefix()
+					dc.SendMessage(ircErr.Message)
+				} else if err != nil {
+					dc.logger.Printf("failed to handle message %q: %v", msg, err)
+					dc.Close()
+				}
+			case eventStop:
+				u.forEachDownstream(func(dc *downstreamConn) {
+					dc.Close()
+				})
+				for _, n := range u.networks {
+					n.stop()
+				}
+				detachTicker.Stop()
+				return
+			default:
+				u.srv.Logger.Printf("received unknown event type: %T", e)
 			}
-		case eventStop:
-			u.forEachDownstream(func(dc *downstreamConn) {
-				dc.Close()
+		case <-detachTicker.C:
```

Comment:
Instead of a ticker, maybe we could have a `time.Timer` set to the next potential detach timeout?

Reply:
We'd need to recompute the next potential detach timeout when a channel detach timeout changes, I'd rather not add this complexity

Reply:
I'm not that comfortable with looping over the channels each 10s for each user.

What about this other plan: add a `*time.Timer` field to `upstreamConn`. Introduce a new `eventDetachChannel` event. If the channel is configured to auto-detach, create the timer like so:

```go
func (uc *upstreamChannel) setAutoDetach(dur time.Duration) {
    if uc.detachTimer != nil {
        uc.detachTimer.Stop()
        uc.detachTimer = nil
    }

    if dur == 0 {
        return
    }

    uc.detachTimer = time.AfterFunc(dur, func() {
        uc.net.user.events <- eventDetachChannel{…}
    })
}
```

This should be simpler than my previous proposal. Wdyt?

Side note, per https://github.com/golang/go/issues/28100, `Timer.C` is unused for timers created via `time.AfterFunc`.

Reply:
Hmmm. I like this better than looping every 10s. We'll have to call this function when joining, parting, receiving messages, and changing channel detach timeouts, but it's worth it